helper/storagepacker/storagepacker_v2.go

```diff
@@ -0,0 +1,774 @@
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to read packed storage entry: {{err}}", err)
+	}
+	if storageEntry == nil {
+		return nil, nil
+	}
+
+	bucket, err := s.DecodeBucket(storageEntry)
+	if err != nil {
+		return nil, err
+	}
+
+	s.bucketsCacheLock.Lock()
+	s.bucketsCache.Insert(cacheKey, bucket)
+	s.bucketsCacheLock.Unlock()
+
+	return bucket, nil
+}
+
+// NOTE: Don't put inserting into the cache here, as that will mess with
+// upgrade cases for the identity store as we want to keep the bucket out of
+// the cache until we actually re-store it.
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	uncompressedData, notCompressed, err := compressutil.Decompress(storageEntry.Value)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decompress packed storage entry: {{err}}", err)
+	}
+	if notCompressed {
+		uncompressedData = storageEntry.Value
+	}
+
+	var bucket Bucket
+	err = proto.Unmarshal(uncompressedData, &bucket)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode packed storage entry: {{err}}", err)
+	}
+
+	cacheKey := s.GetCacheKey(storageEntry.Key)
+	lock := locksutil.LockForKey(s.storageLocks, cacheKey)
+	lb := &LockedBucket{
+		LockEntry: lock,
+		Bucket:    &bucket,
+	}
+	lb.Key = storageEntry.Key
+
+	return lb, nil
+}
+
+// Put stores a packed bucket entry
+func (s *StoragePackerV2) PutBucket(ctx context.Context, bucket *LockedBucket) error {
+	if bucket == nil {
+		return fmt.Errorf("nil bucket entry")
+	}
+
+	if bucket.Key == "" {
+		return fmt.Errorf("missing key")
+	}
+
+	bucket.Lock()
+	defer bucket.Unlock()
+
+	if err := s.storeBucket(ctx, bucket, true); err != nil {
+		return errwrap.Wrapf("failed at high level bucket put: {{err}}", err)
+	}
+
+	s.bucketsCacheLock.Lock()
+	s.bucketsCache.Insert(s.GetCacheKey(bucket.Key), bucket)
+	s.bucketsCacheLock.Unlock()
+
+	return nil
+}
+
+func (s *StoragePackerV2) shardBucket(ctx context.Context, bucket *LockedBucket, cacheKey string, allowLocking bool) error {
+	if allowLocking {
+		s.shardLock.Lock()
+		defer s.shardLock.Unlock()
+
+		bucketLock := bucket.LockEntry
+		defer func() {
+			for lock := range shardLocks {
+				// Let the initial calling function take care of the highest level lock
+				if lock != bucketLock {
+					lock.Unlock()
+				}
+			}
+
+			// Empty the map
+			shardLocks = make(map[*locksutil.LockEntry]struct{}, 32)
+		}()
+	}
+
+	numShards := int(math.Pow(2.0, float64(s.BucketShardBits)))
+
+	// Create the shards and lock them
+	s.Logger.Info("sharding bucket", "bucket_key", bucket.Key, "num_shards", numShards)
+	defer s.Logger.Info("sharding bucket process exited", "bucket_key", bucket.Key)
+
+	shardLocks[bucket.LockEntry] = struct{}{}
+
+	shards := make(map[string]*LockedBucket, numShards)
+	for i := 0; i < numShards; i++ {
+		shardKey := fmt.Sprintf("%x", i)
+		lock := locksutil.LockForKey(s.storageLocks, cacheKey+shardKey)
```

Comment:
This doesn't seem deadlock-safe to me. The locks from s.storageLocks can be acquired in some arbitrary order, correct?  So any guarantee depending on lock ordering goes out the window. Maybe it's safe if no non-sharding thread acquires more than one bucket lock, and we do have the lock to ensure only one sharing operation at a time.

However, I am worried about the following sequence:
A acquires storageLock 1 in PutBucket
B acquired storageLock 2 in PutBucket for a different bucket
A needs to shard, acquires shardLock
B also needs to shard, blocks on shardLock
A's cacheKey+shardKey duplicates to storageLock 2, deadlock?

We acquire storageLocks both before and after shardLock.